common/dm.c

```diff
@@ -21,28 +21,40 @@ device_probe(struct device *dev)
 	assert(dev->drv);
 	assert(dev->drv->probe);
 
+	/* Skip already-probed devices. */
 	if (dev->flags & DEVICE_FLAG_RUNNING)
 		return SUCCESS;
+	if (dev->flags & DEVICE_FLAG_MISSING)
+		return ENODEV;
+
+	/* Probe all devices this device depends on. */
 	if (dev->bus && (err = device_probe(dev->bus)))
 		return err;
 	if (dev->clockdev && (err = device_probe(dev->clockdev)))
 		return err;
 	if (dev->irqdev && (err = device_probe(dev->irqdev)))
 		return err;
-	if ((err = dev->drv->probe(dev)))
+	if (dev->supplydev && (err = device_probe(dev->supplydev)))
 		return err;
 
-	debug("Finished probing device %s", dev->name);
-	dev->flags |= DEVICE_FLAG_RUNNING;
-	return SUCCESS;
+	/* Probe the device itself, and report any errors. */
+	if ((err = dev->drv->probe(dev)) == SUCCESS) {
+		dev->flags |= DEVICE_FLAG_RUNNING;
+		debug("dm: Probed %s", dev->name);
+	} else if (err == ENODEV) {
+		dev->flags |= DEVICE_FLAG_MISSING;
+		warn("dm: Failed to probe %s (missing)", dev->name);
+	} else {
+		panic("dm: Failed to probe %s (%d)", dev->name, err);
```

Comment:
Does this PR assume missing devices are not fatal, but all other errors are fatal? Maybe we could include a device flag for devices where probing failures are non-fatal?

Reply:
The problem here is if we include external devices (i.e. on the I²C bus) in the device list. We don't want to panic if the firmware is run on a board that doesn't have them. The current code works the same way (it's just less clear about it). Probe functions that return `ENODEV` trigger a warning. Any other failure code is fatal. Perhaps, if we expect there to be missing devices, this should still be a debug message?

Reply:
I don't think missing being non-fatal needs a separate flag. Only those probe functions that can return `ENODEV` can fail non-fatally. For example, probe functions for SoC-internal hardware never return `ENODEV`, so they can't be treated as missing. Remember, all other types of failure are fatal.

Reply:
I've changed this to a `debug` print, as missing devices (e.g. external regulators) are to be expected.

Reply:
I am sorry to be so scrupulous, but I prefer this as a `warn` rather than a `debug` statement. Reason being, a `warn` implies this could _potentially_ be a problem, but there are situations where this is okay.

> Only those probe functions that can return `ENODEV` can fail non-fatally.

This was my original question, and this makes sense to me. I can give approval with the change of the `debug` statement to `warn`.

Reply:
No worries, it makes sense. I'll change it back.